RoboJackets/robocup-software
# Basic defense play

## Description
There's a surprising amount of new stuff here:
 * goalie_id from the /referee/goalie_id ROS topic is now a part of WorldState (`world_state.game_info.goalie_id`)
 * intercept action/skill added (this is broken)
 * goalie tactic added
 * basic_defense play added

## Steps to test
FIRST: Change the number of robots to 6. This can be done by clicking on grSim, then in the left menu finding Game > Robots Count and changing the count to 6. (If you can't see the current number, scroll to the right.)

![Screenshot from 2021-06-18 23-00-05](https://user-images.githubusercontent.com/30275369/122629153-f5a08700-d088-11eb-94ae-8a4d6cfa6c6b.png)

### Test Case 1
1. Make sure on any and all runs that robot 0 is the goalie robot (the one in the penalty box)
(This is set off of the `/referee/goalie_id` topic; you can check rqt to confirm that they match.)

### Test Case 2
1. Start sim.
2. Drag ball around field.

Expected result: 
 * Two robots mark the other team, three form a wall, one robot goes to be a goalie. This holds regardless of where the ball is. 
 * Goalie stays between the ball and the center of the goal, moving closer/further depending on the ball's distance, and never going beyond the wall.
 * Non-goalie robots don't enter the goalie box (currently broken)

### Test Case 3
1. Comment out line 17 of `play/basic_defense.py` (to disable the wall).
2. Move other robots out of the way
3. Take a bunch of shots on goal from different angles/speeds.

Expected result:
 * Goalie moves to block shots & settles the ball once the shot has been blocked.
 * Goalie doesn't panic and try to block a shot that isn't coming.

This last test case is the most likely to fail, since I essentially hacked around the non-working intercept motion command with my own move to point command.

### Test Case 4
1. Reenable the wall
2. Simulate playing soccer against the defense

Expected result:
 * No defense-breaking bugs are exposed in normal gameplay conditions.
